CHT tutorials for OpenFOAM and CalculiX

Alphaoo1 opened the pull request:
I have added the [OpenFOAM+CalculiX CHT case](https://precice.discourse.group/t/precice-mesh-id-1-error-openfoam-calculix-cht/313/7), which have some issues from CalculiX side.
I took the general setup of the flap case from the FSI.

I have adapted the CHT part similar to the heat-exchanger case.

I am currently working on making a flow over the heat plate case similar to [OF+Nutils](https://github.com/precice/tutorials/tree/master/CHT/flow-over-plate/buoyantPimpleFoam-nutils). I will push as soon as I am done, working towards Issue #103 .

Alphaoo1 commented:
I have set up a case for flow over hot plate, but in this case as well I am getting a same error as mention [here](https://precice.discourse.group/t/openfoam-calculix-heat-exchanger-cht-tutorial-does-not-run/312).
```
Starting CHT analysis via preCICE...
About to enter preCICE setup in Calculix with names CalculiX and config.yml 
Setting up preCICE participant CalculiX, using config file: config.yml
---[precice]  This is preCICE version 2.1.0
---[precice]  Revision info: v2.1.0-2-g42abc38c
---[precice]  Configuring preCICE with configuration "./precice-config.xml"
---[precice]  I am participant "CalculiX"
Set ID Found 
Read data 'Temperature' found.
---[precice] ERROR:  The given Mesh ID "-1" is unknown to preCICE.
```

Next, I will try and reinstall the CalculiX adapter because I think at least the tried heat exchanger should have worked.

Alphaoo1 commented:
The only thing to edit in the CalculiX part now is the material properties (to be consistent with the [benchmark case](https://github.com/precice/openfoam-adapter/wiki/Tutorial-for-CHT:-Flow-over-a-heated-plate#case-setup)).
```
*MATERIAL, Name=EL
** Material Properties
*ELASTIC
 400000, 0.3
*DENSITY
 3000
*SPECIFIC HEAT
446
*CONDUCTIVITY
50.,0
```

Hopefully, this is of help.

KyleDavisSA commented:
I tried setting up the case as well and got segmentation faults using `Heat-Flux` and `Temperature` data types in the `config.yml` file. Did you experience this? I am going to look into this and open an issue if I find a problem. Perhaps I made a mistake in the setup of the problem.

KyleDavisSA commented:
Is this meant to be for the flow over a plate CHT problem?

Alphaoo1 commented:
>Is this meant to be for the flow over a plate CHT problem?

@KyleDavisSA Yes. There is also one case which is modified FSI flap case.

>I tried setting up the case as well and got segmentation faults using Heat-Flux and Temperature data types in the config.yml file. Did you experience this? I am going to look into this and open an issue if I find a problem. Perhaps I made a mistake in the setup of the problem.

No, I did not face this issue. Seems to be something else than what I am getting.

KyleDavisSA commented:
> Read data 'Temperature' found.
---[precice] ERROR:  The given Mesh ID "-1" is unknown to preCICE.

I got the same error. The `Heat-Flux` needs to be defined on a surface, so it requires the `interface.sur` file to be defined. The temperature on a surface also needs to be given, so the boundary condition `Sink-Temperature` should probably be used here instead of `Temperature`. Could you supply the `interface.sur` file ot check if this is correct?

Alphaoo1 commented:
I was able to make the flow over plate example work, with `Sink-Temperature` and `Heat-Flux`.
I used the default CalculiX adapter(with CCX v2.16) without any modifications.

I was also able to run the Heat exchanger tutorial with the new adapter.
![result-success_t=1](https://user-images.githubusercontent.com/27627601/93325474-b0a8a500-f817-11ea-8867-33ede0a0dd3c.jpeg)
![result-success_t=10](https://user-images.githubusercontent.com/27627601/93325475-b1413b80-f817-11ea-9c3a-c1300641dad2.jpeg)

You may see that the results are a bit weird, but if you see the image below they make sense.
I will try and correct my CacluliX set-up. I think BC is not correctly applied. 
![Screenshot_20200916_120603](https://user-images.githubusercontent.com/27627601/93325500-b900e000-f817-11ea-9c33-555a1c5db36b.png)

MakisH commented:
@Alphaoo1 similar to #92, could you please also remove the result files from here as well?

Alphaoo1 commented:
@MakisH I have cleaned up the cases to the best of my knowledge.
Let me know if I missed something.

I have kept the `.nam`, `.dfl`, `.flm` files unlike the usual tutorials because I haven't figured out how to write a script for exporting the extra files each time we run a case.

:smile:

MakisH commented:
Thank you for cleaning up, @Alphaoo1! We are now down to 226 changed files.

I see changes across multiple tutorials and I am confused on what is exactly the purpose of this PR. Is it to add the flow-over-plate scenario we have with OpenFOAM, but with OpenFOAM and CalculiX? If yes, could you please remove any changes not related to this tutorial?

Is the solver supposed to be `buoyantPimpleFoam` or `chtMultiRegionFoam`?

Alphaoo1 commented:
> Thank you for cleaning up, @Alphaoo1! We are now down to 226 changed files.
> 
> I see changes across multiple tutorials and I am confused on what is exactly the purpose of this PR. Is it to add the flow-over-plate scenario we have with OpenFOAM, but with OpenFOAM and CalculiX? If yes, could you please remove any changes not related to this tutorial?
> 

This PR is for flow over plate with OpenFOAM and CalculiX.
I removed the extra one I modified case I had in CHT. There should be 38 files from my case. I don't know why there are files popping up. I even tried to clone the current develop directory and add my case to it, but still, there are files that I did not modify. And I am a bit reluctant to touch these files since somebody else must have changed them and I am not sure what to keep and what to delete.
Some tips are welcome here. :sweat_smile: 

> Is the solver supposed to be `buoyantPimpleFoam` or `chtMultiRegionFoam`?

The solver is `buoyantPimpleFoam`. I copied the Fluid part from the nutils directory.
I did not change anything other than the preCICEdict as far as I remember.

MakisH commented:
The additional files were there because your `develop` was a bit behind the `precice:develop`. I updated your branch and now there are 37 files, nice! :smile: I will now look into details.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/runSolid:
```diff
@@ -0,0 +1,14 @@
+echo "Running the Solid participant..."
+
+# Run
+ccx_preCICEv216 -i Solid/Solid_plate -precice-participant Calculix #> Solid.log
```
For consistency:
- We use CalculiX everywhere as `ccx_preCICE`. If a specific version is needed (I guess not), better describe it in the README.
- We need that the tutorial always produces the `Solid.log` and `Fluid.log` files.

```suggestion
ccx_preCICE -i Solid/Solid_plate -precice-participant Calculix > Solid.log 2>&1
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/precice-config.xml:
```diff
@@ -0,0 +1,66 @@
+        <read-data name="Heat-Flux" mesh="OpenFOAM-Mesh"/>
+        <write-data name="Sink-Temperature-0" mesh="OpenFOAM-Mesh"/>
+        <mapping:nearest-neighbor direction="read" from="Calculix_Mesh" to="OpenFOAM-Mesh" constraint="consistent" timing="initial"/>
+        <export:vtk directory="preCICE-OF-output" />
```
In other tutorials we don't export VTKs, so let's remove this line:
```suggestion
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/precice-config.xml:
```diff
@@ -0,0 +1,66 @@
+        <use-data name="Heat-Flux"/>
+    </mesh>
+
+    <mesh name="Calculix_Mesh">
```
Inconsistency in the name style: `OpenFOAM-Mesh` vs `Calculix_Mesh`.

I suggest `OpenFOAM-Mesh` and `CalculiX-Mesh`.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/precice-config.xml:
```diff
@@ -0,0 +1,66 @@
+        <read-data name="Sink-Temperature-0" mesh="Calculix_Mesh"/>
+        <write-data name="Heat-Flux" mesh="Calculix_Mesh"/>
+        <mapping:nearest-neighbor direction="read" from="OpenFOAM-Mesh" to="Calculix_Mesh" constraint="consistent" timing="initial"/>
+        <export:vtk directory="preCICE-CCX-output" />
```
Also remove:
```suggestion
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/precice-config.xml:
```diff
@@ -0,0 +1,66 @@
+
+    <coupling-scheme:serial-implicit>
+        <time-window-size value="0.01"/>
+        <max-time value="10"/>
```
We want that tutorials are rather quick, at least for a first run. Better reduce the time to 1, also for consistency with OF-OF:
```suggestion
        <max-time value="1"/>
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/Allrun:
```diff
@@ -0,0 +1,35 @@
+cd ..
+
+# Participant 2: Nutils
+Participant2="Nutils"
```
```suggestion
Participant2="Solid"
```

MakisH reviewed (changes requested):
I have a few comments mostly for consistency with other tutorials, having another look at which CalculiX files we can remove (try moving them and running the case), and a few more important suggestions for the run scripts.

Keep up the hard work, we are getting there! :+1: :+1:

Alphaoo1 commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/precice-config.xml:
```diff
@@ -0,0 +1,66 @@
+
+    <solver-interface dimensions="3">
+
+    <data:scalar name="Sink-Temperature-0"/>
+    <data:scalar name="Heat-Flux"/>
```
In CalculiX adapter, the [keyword](https://github.com/precice/calculix-adapter/blob/ad446eb52b3c889f61268755db8dba75f2bf1d83/adapter/PreciceInterface.c#L612-L623) for Heat flux is `Heat-Flux` and for sink temperature is `Sink-Temperature-`

When I use `Heat-Flux-0` instead of `Heat-Flux`, I get an error:
```
`ERROR: Write data 'Heat-Flux-0' does not exist!
```

MakisH replied:
But if you use `Heat-Flux-0` everywhere (also in the OpenFOAM and CalculiX adapters), then it should work. The OpenFOAM adapter supports this.

Alphaoo1 commented:
@MakisH In a nutshell I made the following changes:

- Updated the `Allrun` and `Allclean` scripts.
- Updated the Readme.
- Updated precice-config.xml
- Removed the headers from Fluid participant.
- Removed unecessary files from Solid directory.

As for the Solid participant and it's files, I have added things one can do to generate those files. 
This may be used as a backup options if some file gets deleted.
But as for a new comer, or someone who is new to CalculiX I think these might be complicated.
I myself am not aware of how to write the `pre-process.fbd` file which creates all the files (similar to what we have in FSI tutorials).
I will let you know when I am able to do that, o may be you can mention someone with expertise in CalculiX.
Looking forward to your comments and suggestions.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/Allrun:
```diff
@@ -1,35 +1,120 @@
+
+    # Workaround for issue #26
+    echo "Problems with time directories without results? Run the script removeObsoleteFolders.sh and see issue #26 on GitHub."
+    # ./removeObsoleteFolders.sh
```
We need to keep this script to clean empty result directories such as `0.19/` (which would appear between `0.1` and `0.2` in openfoam.com). We also need to add this script (copy it from another tutorial and check the directory names inside it).

```suggestion
    ./removeObsoleteFolders.sh
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/Allrun:
```diff
@@ -1,35 +1,120 @@
+    # Note: ".foam" triggers the native OpenFOAM reader of ParaView.
+    # Change to ".OpenFOAM" to use the OpenFOAM reader provided with OpenFOAM.
 fi
+
+echo ""
+echo "### NOTE ### Make sure to use the correct solver for your OpenFOAM version! (pimpleFoam for OpenFOAM v1806, OpenFOAM 6, or newer, vs pimpleDyMFoam for older) You may change this in your Fluid/system/controlDict file, if needed."
```
We don't need these lines in this case, as we are using buoyantPimpleFoam (and don't have moving meshes):
```suggestion
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/precice-config.xml:
```diff
@@ -4,8 +4,8 @@
         <sink filter="%Severity% > debug and %Rank% = 0" format="---[precice] %ColorizedSeverity% %Message%" enabled="true"/>
-        <sink type="file" output="debug_OpenFOAM.log" filter="%Participant% = OpenFOAM" />
-        <sink type="file" output="debug_Calculix_Mesh.log" filter="%Participant% = Calculix" />
+        <!--<sink type="file" output="debug_OpenFOAM.log" filter="%Participant% = OpenFOAM" />
+        <sink type="file" output="debug_Calculix_Mesh.log" filter="%Participant% = Calculix" />-->
```
I am not sure if we should keep such commented-out lines. One the one hand, they help the user discover features, but on the other hand we should do this consistently across tutorials and also avoid increasing the size the config file. In any case, such "dead code" is usually easy to miss when updating things.

I would vote for removing it. @uekerman what do you think?

```suggestion
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/runFluid:
```diff
@@ -7,7 +7,12 @@
+        cp -r 0.orig 0
+    cd ..
 # Prepare
 blockMesh -case Fluid
```
Here, you can write this in a simpler way:

```suggestion
Participant1="Fluid"

# Prepare
cp -r Fluid/0.rig Fluid/0
blockMesh -case Fluid
```

In a second view, I would even say that we don't need this additional step in this tutorial. We only need this when we apply pre-processing steps that directly change the files in the `0/` directory (e.g. when we initialize with a potential flow).

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/Allrun:
```diff
@@ -1,35 +1,120 @@
+    echo "Preparing the ${Participant1} participant..."
+
+    cd $Participant1 
+    echo "  Restoring 0/ from 0.orig/..."
+        rm -rfv 0
+        cp -r 0.orig 0
```
We actually don't need this step here, right? We don't have any step that directly modifies the `0/` directory, so we don't need a backup. You can rename the `0.orig/` to `0/` and remove this step from `Allrun` and `runFluid`.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/README.md:
```diff
@@ -0,0 +1,85 @@
+
+## Case setup: flow over heated plate
+
+The setup for this tutorial is identical to the [flow over heated plate using OpenFOAM](https://github.com/precice/openfoam-adapter/wiki/Tutorial-for-CHT:-Flow-over-a-heated-plate#case-setup) with the only exception that the width of the scenario in z-direction is `0.05` instead of `0.4`. 
```
Why is that?

Alphaoo1 replied:
I started from the Nutils-OF CHT tutorial which was using z-direction width as `0.05`. 
Please note that I used the OpenFOAM part as is form the [Nutils-OF flow over plate](https://github.com/precice/tutorials/tree/master/CHT/flow-over-plate/buoyantPimpleFoam-nutils#case-setup-flow-over-heated-plate) which was initially suggested to me. 

I am not sure why this was initially done for Nutils-OF CHT tutorial.

MakisH replied:
This is ok, don't worry! It is maybe better than the OpenFOAM case: it makes the cells more "square", improving the stability. I was just wondering if you had a specific reason in mind.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/README.md:
```diff
@@ -0,0 +1,85 @@
+For running this tutorial, you further have to install
+
+* **preCICE**, see [preCICE wiki](https://github.com/precice/precice/wiki/Building).
+* **Python bindings**, see [`precice/python-bindings`](https://github.com/precice/python-bindings)
```
This should not be needed:
```suggestion
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/README.md:
```diff
@@ -0,0 +1,85 @@
+* **Python bindings**, see [`precice/python-bindings`](https://github.com/precice/python-bindings)
+* **OpenFOAM**, see [Notes on OpenFOAM](https://github.com/precice/openfoam-adapter/wiki/Notes-on-OpenFOAM).
+* **OpenFOAM adapter**, see [OpenFOAM adapter wiki](https://github.com/precice/openfoam-adapter/wiki/Building). If you have problems compiling, see the [troubleshooting section](https://github.com/precice/precice/wiki/CHT-with-OpenFOAM-and-FEniCS#troubleshooting) below.
+* **CalculiX adapter, see [CalculiX adapter wiki](https://github.com/precice/calculix-adapter/wiki)
```
```suggestion
* **CalculiX adapter**, see [CalculiX adapter wiki](https://github.com/precice/calculix-adapter/wiki)
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/README.md:
```diff
@@ -0,0 +1,85 @@
+* **OpenFOAM and OpenFOAM adapter:** To make sure that everything is working properly, you should run the [similar OpenFOAM-OpenFOAM tutorial case](https://github.com/precice/openfoam-adapter/wiki/Tutorial-for-CHT:-Flow-over-a-heated-plate).
+
+### Generating CacluliX files
+After installing the CalculiX adapter, CalculiX on it's own should also be available in the system. If not, an easy to use precomplied [tool](http://www.calculixforwin.com/) may be used to produced the required files (the commands remain same, use the pre-processing option).
```
This is actually not correct. It is not necessary to compile a "normal" CalculiX to compile `ccx_preCICE`.

Alphaoo1 replied:
I think may be my wording was not clear, I was trying to say the same, we do not need to install "normal" CalculiX.
Isn't it installed directly when we installed the adapter?

MakisH replied:
If one runs `make` only in the adapter's repository, no. The user only needs to have the CalculiX code somewhere. But I agree that a CalculiX user would also install the normal `ccx`.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/README.md:
```diff
@@ -0,0 +1,85 @@
+* **OpenFOAM and OpenFOAM adapter:** To make sure that everything is working properly, you should run the [similar OpenFOAM-OpenFOAM tutorial case](https://github.com/precice/openfoam-adapter/wiki/Tutorial-for-CHT:-Flow-over-a-heated-plate).
+
+### Generating CacluliX files
+After installing the CalculiX adapter, CalculiX on it's own should also be available in the system. If not, an easy to use precomplied [tool](http://www.calculixforwin.com/) may be used to produced the required files (the commands remain same, use the pre-processing option).
```
Typo:

```suggestion
After installing the CalculiX adapter, CalculiX on it's own should also be available in the system. If not, an easy to use precomplied [tool](http://www.calculixforwin.com/) may be used to produce the required files (the commands remain same, use the pre-processing option).
```

Not sure yet about the CalculiX Launcher, but I will have a look when I run the case.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/Allrun:
```diff
@@ -1,35 +1,120 @@
+Participant2="Solid"
+Solver2="ccx_preCICEv216"
+
+    # You can use CalculiX CGX to setup the structural simulation from sratch.
+    # This will re-generate the all.msh, fix1_beam.nam, interface_beam.nam files.
+    #
+    # # Prepare in silent mode
+    # echo "Preparing the ${Participant2} participant..."
+    # cd ${Participant2}
+    #     echo "  Executing cgx (provided by CalculiX, make sure this exists)..."
+    #     cgx -bg pre_flap.fbd > prepare_flap.log 2>&1
+    # cd ..
```
In the README it looks like this is the respective step:
```
cgx -c Mesh_Coarse_OUT.inp
```

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/README.md:
````diff
@@ -0,0 +1,85 @@
+prnt se
+```
+
+* To generate the mesh and node files, the following commands are used:
````
To be consistent with the rest of the README:

```suggestion
* To generate the mesh and node files:
```

you could also add a `:` at the end of the next bullets.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/Fluid/runFluid:
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
```
Still needs to be deleted.

MakisH reviewed (changes requested):
Thank you for applying most of the suggestions already! Here are a few more ones (in most of them you can just apply the diff suggestions).

> As for the Solid participant and it's files, I have added things one can do to generate those files.
This may be used as a backup options if some file gets deleted.
But as for a new comer, or someone who is new to CalculiX I think these might be complicated.

I agree that we should not assume previous CalculiX knowledge and I like that you added the steps in the README! :smile: 

I have not yet tried to run the tutorial, I will try after we resolve most of these points.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/runFluid:
```diff
@@ -1,18 +1,10 @@
-#!/bin/bash
```
This is good, keep it! It makes it explicit that this is a Bash script.

MakisH replied:
I meant to delete the file `./Fluid/runFluid`, since we also have the `./runFluid`

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/runSolid:
```diff
@@ -1,5 +1,3 @@
-#!/bin/bash
```
Same here: do not remove this line.

MakisH commented on CHT/flow-over-plate/buoyantPimpleFoam-CalculiX/dummy:
```diff
@@ -0,0 +1,3 @@
+ nactdog
+ 
+ nacteq
```
What is this file for?

MakisH commented:
@Alphaoo1 we restructured the tutorials as a larger project https://github.com/orgs/precice/projects/5

This contribution is still quite important, but I think it woul dbe much easier to close this now and submit a new PR again, with the new structure in mind. Since now most files are already there (OpenFOAM case, README.md, scripts, `precice-config.xml`), you can just add a new `solid-calculix` case in the `flow-over-heated-plate` directory. You can also be the first one to try out the new [contributing guidelines](https://precice.org/community-contribute-to-precice.html) and see if they help.

Thank you once more already for the work so far, I am only closing this to keep a better overview of what is actively open.

Alphaoo1 commented:
@MakisH Yes, I will surely have a look at the contributions guidelines and see how can I contribute.
Currently busy with the comparisons of the XDEM+OF simulations for the proposal by Benjamin :wink:
